tests/pubsub/testpreamblestore.nim

```diff
@@ -0,0 +1,157 @@
+import unittest2
+
+{.used.}
+
+import chronos
+import
+  ../../libp2p/[
+    peerid,
+    protocols/pubsub/gossipsub/preamblestore,
+    protocols/pubsub/gossipsub/types,
+    protocols/pubsub/rpc/message,
+    protocols/pubsub/pubsubpeer,
+    protocols/pubsub/rpc/messages,
+  ]
+import ../utils/async_tests
+import ./utils
+import ../helpers
+
+proc mockPreamble(
+    id: MessageId,
+    sender: PubSubPeer,
+    startOffset: Duration = 0.seconds,
+    expireOffset: Duration = 10.seconds,
+): PreambleInfo =
+  let now = Moment.now()
+  PreambleInfo.init(
+    ControlPreamble(messageID: id, messageLength: 123, topicID: "test"),
+    sender = sender,
+    startsAt = now + startOffset,
+    expiresAt = now + expireOffset,
+  )
+
+suite "preamble store":
+  teardown:
+    checkTrackers()
+
+  const topic = "foobar"
+
+  asyncTest "insert and retrieve":
+    var store = PreambleStore.init()
+    let peer = PubSubPeer.new(randomPeerId(), nil, nil, GossipSubCodec_12, 0)
+    let m1 = MessageId(@[1.byte, 1.byte, 1.byte])
+    let info = mockPreamble(m1, peer, 0.seconds, 5.seconds)
+
+    store.insert(m1, info)
+    check:
+      store.hasKey(m1)
+      store[m1] == info
+
+  asyncTest "heap order is preserved on inserts":
+    var store = PreambleStore.init()
+    for i in 0 .. 9:
+      let id = MessageId(@[i.byte])
+      let peer = PubSubPeer.new(randomPeerId(), nil, nil, GossipSubCodec_12, 0)
+      let p = mockPreamble(id, peer, 0.seconds, i.seconds)
+      store.insert(id, p)
+
+    var last = Moment.low()
+    while true:
+      let popped = store.popExpired(Moment.now() + 20.seconds)
+      if popped.isNone:
+        break
+      check popped.get().expiresAt >= last
+      last = popped.get().expiresAt
+
+  asyncTest "deletion marks AND heap pop discards deleted":
+    var store = PreambleStore.init()
+
+    let m1 = MessageId(@[1.byte, 1.byte, 1.byte])
+    let m2 = MessageId(@[2.byte, 2.byte, 2.byte])
+
+    let peer = PubSubPeer.new(randomPeerId(), nil, nil, GossipSubCodec_12, 0)
+    let p1 = mockPreamble(m1, peer, 0.seconds, 1.seconds)
+    let p2 = mockPreamble(m2, peer, 0.seconds, 2.seconds)
+    store.insert(m1, p1)
+    store.insert(m2, p2)
+    store.del(m1)
+    # p1 is now deleted in heap, but still there
+    let res = store.popExpired(Moment.now() + 5.seconds)
+    check:
+      res.isSome()
+      res.get().messageId == m2 # p1 was skipped
+
+  asyncTest "overwrite twice keeps latest, heap has garbage record":
+    var store = PreambleStore.init()
```

Comment:
Maybe make `store` a `{.threadVar.}` and initialize it in an `asyncSetup` block?